src/postgres-util/src/lib.rs

```diff
@@ -309,50 +317,88 @@ impl Config {
                 port,
                 user,
-                public_key,
+                public_key: _,
                 private_key,
             } => {
-                let tcp_stream = TokioTcpStream::connect((host.as_str(), *port)).await?;
-                let mut session = AsyncSession::new(tcp_stream, Some(SessionConfiguration::new()))?;
-                session.handshake().await?;
+                let tempdir = tempfile::Builder::new()
+                    .prefix("ssh-tunnel-key")
+                    .tempdir()?;
+                let path = tempdir.path().join("key");
+                let mut tempfile = File::create(&path)?;
+                tempfile.set_permissions(std::fs::Permissions::from_mode(0o600))?;
+                tempfile.write_all(private_key.as_bytes())?;
+                tempfile.set_permissions(std::fs::Permissions::from_mode(0o400))?;
 
-                session
-                    .userauth_pubkey_memory(user, Some(public_key), private_key, None)
-                    .await?;
+                let mut builder = openssh::SessionBuilder::default();
+                builder.user(user.clone()).port(*port).keyfile(&path);
+                let session = builder.connect(host).await?;
 
-                if !session.authenticated() {
-                    bail!("failed SSH authentication")
-                };
+                // Delete the private key for safety: since `ssh` still has an open handle to it,
+                // it still has access to the key.
+                drop(tempfile);
+                std::fs::remove_file(&path)?;
+                drop(tempdir);
 
-                // Advertise the source connection to Postgres as coming from the bastion host
-                let src = Some((host.as_str(), *port));
+                // Ensure session is healthy
+                session.check().await?;
+
+                // Loop trying to find an open port
+                let local_port = loop {
+                    let mut rng: rand::rngs::StdRng = rand::SeedableRng::from_entropy();
+                    // Choosing a dynamic port according to RFC 6335
+                    let local_port: u16 = rng.gen_range(49152..65535);
+
+                    let local = openssh::Socket::new(&("localhost", local_port))?;
+                    let remote = (self.host.as_str(), self.port);
+                    let remote = openssh::Socket::new(&remote)?;
+
+                    match session
+                        .request_port_forward(openssh::ForwardType::Local, local, remote)
+                        .await
+                    {
+                        Err(err) => match err {
+                            openssh::Error::Ssh(err)
+                                if err.to_string().contains("forwarding request failed") =>
+                            {
+                                tracing::warn!(
+                                    "Port {local_port} already in use, testing another port"
+                                );
```

Comment:
not too bad! maybe add a `TODO` to give up after like...50 tries!

Reply:
:+1:. Originally I kept a `HashMap` of all current open SSH tunnels + attempted ports, but randomly checking ports works just as well and is a lot simpler to manage.

Reply:
perfect!